lib/Sema/TypeCheckType.cpp

```diff
@@ -1640,6 +1640,62 @@ Type TypeChecker::resolveIdentifierType(
   return result;
 }
 
+/// Validate whether type associated with @uatoclosure attribute is correct,
+/// it supposed to be a function type with no parameters.
+/// \returns true if there was an error, false otherwise.
+static bool validateAutoClosureAttr(TypeChecker &TC, const SourceLoc &loc,
+                                    Type paramType) {
+  if (auto *fnType = paramType->getAs<FunctionType>()) {
+    if (fnType->getNumParams() != 0) {
+      TC.diagnose(loc, diag::autoclosure_function_input_nonunit);
+      return true;
+    }
+    // A function type with no parameters.
+    return false;
+  }
+
+  TC.diagnose(loc, diag::autoclosure_function_type);
+  return true;
+}
+
+/// Check whether the type associated with particular source location
+/// has `@autoclosure` attribute, and if so, validate that such use is correct.
+/// \returns true if there was an error, false otherwise.
+static bool validateAutoClosureAttributeUse(TypeChecker &TC, const TypeLoc &loc,
+                                            Type type,
+                                            TypeResolutionOptions options) {
+  auto *TR = loc.getTypeRepr();
+  if (!TR || TR->isInvalid())
+    return false;
+
+  // If is a parameter declaration marked as @autoclosure.
+  if (options.is(TypeResolverContext::FunctionInput)) {
+    if (auto *ATR = dyn_cast<AttributedTypeRepr>(TR)) {
+      const auto attrLoc = ATR->getAttrs().getLoc(TAK_autoclosure);
+      if (attrLoc.isValid())
+        return validateAutoClosureAttr(TC, attrLoc, type);
+    }
+  }
+
+  // Otherwise, let's dig into the type and see if there are any
+  // functions with parameters marked as @autoclosure,
+  // such would be a part of expressions like:
+  // `let _: (@autoclosure () -> Int) -> Void = ...`.
+  bool isValid = true;
+  type.visit([&](Type subType) {
+    if (auto *fnType = subType->getAs<FunctionType>()) {
+      isValid &= llvm::none_of(
```

Comment:
This `&=` is going to short-circuit, meaning we won't see all of the `@autoclosure`-related diagnostics if there are several in a type. Perhaps use an `if` here?

Reply:
I'm not sure what you mean by short-circuit, I picked llvm::none_of intentionally so all of the parameters are checked, and `&=` is there to make sure that if there is at least one failure we return it from this method, but diagnostics are going to be provided for every invalid case.

Reply:
If there were two function types in `type` that had `@autoclosure-related` errors, the second wouldn't get diagnosed because `isValid` would be `true` from the first one. That's all... it's not a big deal.

Reply:
I don't understand, but `isValid` not checked anywhere besides in `return`...

Reply:
Oh, silly me. `&=` doesn't short-circuit and `&&=` doesn't exist. Sorry :(